src/dataflow/analysis/flow-graph.ts

```diff
@@ -117,33 +150,90 @@ function* allInputPaths(startEdge: Edge): Iterable<BackwardsPath> {
   }
 }
 
-/** Returns true if the given check passes for the given path. */
-function evaluateCheckForPath(check: Check, path: BackwardsPath): boolean {
-  // TODO: Support boolean expression trees properly! Currently we only deal with a single string of OR'd conditions.
-  const conditions: CheckCondition[] = [];
-  switch (check.expression.type) {
-    case 'and':
-      throw new Error(`Boolean expressions with 'and' are not supported yet.`);
-    case 'or':
-      for (const child of check.expression.children) {
-        assert(child.type !== 'or' && child.type !== 'and', 'Nested boolean expressions are not supported yet.');
-        conditions.push(child as CheckCondition);
+/**
+ * Collects all the tag claims along the given path, canceling tag claims that are
+ * negated by "not" claims for the same tag downstream in the path, and ignoring
+ * "not" claims without corresponding positive claims upstream, as these dangling
+ * "not" claims are irrelevant for the given path. Note that "derives from" claims
+ * only prune paths, and are dealt with during path generation, so they are ignored.
+ */
+function computeTagClaimsInPath(path: BackwardsPath): ClaimIsTag[] {
+  const claims: ClaimIsTag[] = [];
+  // We traverse the path in the forward direction, so we can cancel correctly.
+  const edgesInPath = path.edges.slice().reverse();
+  edgesInPath.forEach(e => {
+    if (!e.claim) {
+      return;
+    }
+    if (e.claim.type === ClaimType.DerivesFrom) {
+      return;
+    } 
+    const tagClaim = e.claim as ClaimIsTag;
```

Comment:
With the above inversion of the if-statement, the compiler should be able to automatically figure out that this is a ClaimIsTag, so you can remove the cast here.

Reply:
Awesome. Done.